src/shogun/lib/parameter_observers/ObservedValue.h

```diff
@@ -1,182 +1,95 @@
+ *
+ * Authors: Giovanni De Toni
+ *
+ */
 
 #ifndef SHOGUN_OBSERVEDVALUE_H
 #define SHOGUN_OBSERVEDVALUE_H
 
 #include <chrono>
-#include <shogun/lib/any.h>
-#include <shogun/lib/common.h>
 #include <utility>
 
+#include <shogun/base/SGObject.h>
+#include <shogun/base/some.h>
+
 /**
  * Definitions of basic object with are needed by the Parameter
  * Observer architecture.
  */
 namespace shogun
 {
+
+	template <class T>
+	class ObservedValueTemplated;
+
 	/* Timepoint */
 	typedef std::chrono::steady_clock::time_point time_point;
 
-	/* Type of the observed value */
-	enum SG_OBS_VALUE_TYPE
-	{
-		TENSORBOARD,
-		CROSSVALIDATION
-	};
-
 	/**
 	 * Observed value which is emitted by algorithms.
 	 */
-	class ObservedValue
+	class ObservedValue : public CSGObject
 	{
 	public:
 		/**
 		 * Constructor
 		 * @param step step
-		 * @param name param's name
-		 * @param value Any-wrapped value of the param
-		 */
-		ObservedValue(
-		    int64_t step, std::string& name, Any value, SG_OBS_VALUE_TYPE type)
-		    : m_step(step), m_name(name), m_value(value), m_type(type)
-		{
-		}
-
-		~ObservedValue(){};
-
-		/**
-		 * Get the step
-		 * @return an integer representing the step
-		 */
-		int64_t get_step() const
-		{
-			return m_step;
-		}
-
-		/**
-		 * Set the step
-		 * @param step step
-		 */
-		void set_step(int64_t step)
-		{
-			m_step = step;
-		}
-
-		/**
-		 * Get the param's name
-		 * @return param's name
-		 */
-		const std::string& get_name() const
-		{
-			return m_name;
-		}
-
-		/**
-		 * Set the param's name
-		 * @param name
-		 */
-		void set_name(const std::string& name)
-		{
-			m_name = name;
-		}
-
-		/**
-		 * Get the Any-wrapped value
-		 * @return Any-wrapped value
+		 * @param name name of the observed value
 		 */
-		const Any& get_value() const
-		{
-			return m_value;
-		}
+		ObservedValue(int64_t step, std::string name);
 
 		/**
-		 * Set the param's value
-		 * @param value
+		 * Destructor
 		 */
-		void set_value(const Any& value)
-		{
-			m_value = value;
-		}
+		~ObservedValue(){};
 
 		/**
-		 * Get the type of this ObservedValue
-		 * @return observed value type
-		 */
-		SG_OBS_VALUE_TYPE get_type() const
+		* Helper method to generate an ObservedValue.
+		* @param step the step
+		* @param name the param's name we are observing
+		* @param value the param's value
+		* @return an ObservedValue object initialized
+		*/
+		template <class T>
+		static Some<ObservedValue>
+		make_observation(int64_t step, std::string name, T value)
 		{
-			return m_type;
+			return Some<ObservedValue>::from_raw(
+			    new ObservedValueTemplated<T>(step, name, value));
 		}
 
+#ifndef SWIG
 		/**
-		 * Set the observed value type
-		 * @param type type of this observed value
-		 */
-		void set_type(const SG_OBS_VALUE_TYPE type)
+		* Return a any version of the stored type.
+		* @return the any value.
+		*/
+		virtual Any get_any()
 		{
-			m_type = type;
+			return m_any_value;
```

Comment:
return value or reference?

Reply:
I guess that return by reference would be okay.